src/lib/raft/raft.c

```diff
@@ -206,10 +206,51 @@ raft_scores_str(const struct raft *raft)
 	return buf;
 }
 
+static inline bool
+raft_leader_is_seen(const struct raft *raft)
+{
+	bool seen = bit_test(&raft->leader_witness_map, raft->self);
+	assert(!seen || raft->leader != 0);
+	return seen;
+}
+
 /** Schedule broadcast of the complete Raft state to all the followers. */
 static void
 raft_schedule_broadcast(struct raft *raft);
 
+/**
+ * Update all leader-related fields on leader appearance and let others know the
+ * leader is seen.
+ */
+static inline void
+raft_set_leader(struct raft *raft, uint32_t leader)
+{
+	assert(raft->leader == 0);
+	assert(leader != 0);
+	raft->leader = leader;
+	bool changed = raft_notify_leader_seen(raft, true, raft->self);
```

Comment:
I don't get it. Do you want to rename `raft_set_leader` or maybe move `raft_notify_leader_seen` outside?
`raft_set_leader` shares common parts of `raft_sm_follow_leader` and `raft_sm_become_leader`, and isn't called anywhere else.

> raft_notify_leader_seen() is called there before raft_set_leader() anyway

It's first called for `source`, so, notify, that `source` sees leader. And in `raft_set_leader` it's called for `self`. So that we remember that we have a direct connection to the leader.

Reply:
> I don't get it. Do you want to rename raft_set_leader or maybe move raft_notify_leader_seen outside?

The problem is not with the name, but with the fact that we set `==` between the facts "leader became X" and "we see the leader directly". It is not true if the leader ID was learned indirectly, not from the leader itself. It can't happen now, but since we will start sending leader ID with each message, it might become the case very soon. When we set leader ID after learning it from a third node, but don't see the leader ourselves yet.